src/schoology-extractor/schoology_extractor/usage_analytics_facade.py

```diff
@@ -3,34 +3,75 @@
 # The Ed-Fi Alliance licenses this file to you under the Apache License, Version 2.0.
 # See the LICENSE and NOTICES files in the project root for more information.
 
+from dataclasses import dataclass
 import logging
+import os
 
 import pandas as pd
+import sqlalchemy
 
+from .helpers import csv_reader
+from .helpers import sync
 from .mapping import usage_analytics as usageMap
 
+logger = logging.getLogger(__name__)
 
-def get_system_activities(usage_df: pd.DataFrame) -> pd.DataFrame:
-
-    # TODO: instead of accepting the usage data frame as an argument, accept the
-    # input file directory as an argument. Get all file names from the input
-    # directory. For each file, look in SQLite to see if the file has been
-    # processed (this will necessitate creating a new table, of course). Then
-    # for each file, use the csv_reader class to read in a DataFrame. If
-    # there are multiple files then combine them into the same DataFrame.
-    # Psuedocode below might be improved with some logging. Logging here
-    # is probably better than logging inside the csv_reader or usageMap.
-    #
-    # def get_system_activities(input_directory) -> pd.DataFrame.
-    #   output = pd.DataFrame()
-    #   for file in os.scandir({only .csv or .csv.gz files}):
-    #       raw_data = csv_reader.load_data_frame(file)
-    #       mapped_data = usageMap.map_to_udm(raw_data)
-    #       output.append(mapped_data)
-    #
-    #   # If reports have overlapping dates then we'll have duplicates
-    #   output.drop_duplicates(inplace=True)
-    #
-    #   return output
-
-    return usageMap.map_to_udm(usage_df)
+
+@dataclass
+class UsageAnalyticsFacade:
+    """
+    Provides business logic to orchestrate processing of data from the LMS(System
+    Activities report CSV) and reshape it into the Ed-Fi LMS Unified Data Model
+    (UDM).
+
+    Parameters
+    ----------
+    db_engine : sqlalchemy.engine.base.Engine
+        Database connectivity for sync process
+    """
+    db_engine: sqlalchemy.engine.base.Engine
+
+    @property
+    def _db_engine(self):
+        assert isinstance(self.db_engine, sqlalchemy.engine.base.Engine)
+        return self.db_engine
+
+    def get_system_activities(self, usage_input_dir: str) -> pd.DataFrame:

```

Comment:
I was thinking about this module being purely functional, rather than being a class. Since there is only one input argument to the class (the `db_engine`) and there is really no need to preserve state, it would be simple to just pass that argument in as an argument `get_system_activities`.

Reply:
It is easy to make it functional. I just wrote the function to have similar apis for the two facades. I can change it if needed.

Reply:
Let's prefer functional at this point, now that we're getting used to that paradigm.

Reply:
✔️